v3/docs/03-runtime/p-smart-contracts/a-overview/index.mdx

```diff
@@ -233,19 +236,28 @@ reached before the contract execution completes, the transaction fails, contract
 reverted, and the gas fee is **not** returned to the user. If the contract execution completes with
 remaining gas, it is returned to the user at the end of the transaction.
 
-The Contracts pallet determines the gas price, which is a conversion between the Substrate
-[**weight system**](/v3/concepts/weight) and a single unit of gas. Thus, to execute a transaction, a user 
-must have a free balance of at least `gas price` \* `gas limit` which can be spent.
+The concept of gas is tightly integrated with substrates [**weight system**](/v3/concepts/weight).
+In fact there is no difference between gas and weight. The specified `gas_limit` directly influences
+the weight of the submitted transaction. The contracts pallet uses
+[post dispatch weight correction](/v3/concepts/weight#post-dispatch-weight-correction) to change
+the pre dispatch weight derived from the `gas_limit` to the actual weight consumed.
+Thus, to execute a transaction, a user  must have a free balance of at least `weight price` \* `gas limit`
+which can be spent. The charged fee however, is based upon the actually consumed gas. The weight price
+is determined due do the usual [transaction fee mechanism](/v3/runtime/weights-and-fees).
```

Comment:
```suggestion
The concept of gas is inextricably related to Substrates [**weight system**](/v3/concepts/weight). The specified `gas_limit` directly influences 
the weight of the submitted transaction. The contracts pallet will take the pre-dispatch weight derived from the `gas_limit`, adjust it according to [post-dispatch weight correction](/v3/concepts/weight#post-dispatch-weight-correction), and determine the actual consumed weight.
Thus, to execute a transaction, a user must have a free balance of at least `weight * gas limit` which can be spent. The charged fee however, is based upon the actually consumed gas. 
The weight price is determined by the usual [transaction fee mechanism](/v3/runtime/weights-and-fees).
```

Reply:
This sounds like the contracts pallet somehow inspects the post-dispatch weight correction in order to determine the gas consumed. In fact it is the other way around. The pallet uses gas metering to determine the consumed weight (remember gas = weight) and uses post-dispatch weight correction to tell the rest of substrate the result of that metering.

Reply:
I see. Let's use your original wordings then.